Zend/zend_alloc.c

```diff
@@ -1904,6 +1967,27 @@ static void zend_mm_free_huge(zend_mm_heap *heap, void *ptr ZEND_FILE_LINE_DC ZE
 /* Initialization */
 /******************/
 
+static zend_result zend_mm_refresh_key(zend_mm_heap *heap)
+{
+	php_random_result result = php_random_algo_xoshiro256starstar.generate(&heap->random_state);
+	ZEND_ASSERT(result.size == sizeof(uint64_t));
```

Comment:
`php_random_result.size` is defined as an `uint64_t`, why this assert?

Reply:
I agree that the `ZEND_ASSERT` is not necessary here: xoshiro256** is a 64-bit engine and that will not change.

As for the question: You probably misread the assertion. It's not asserting that size of `result.size` matches that of `uint64_t`, but rather that the *contents* of it match a `uint64_t`. The `result.size` field indicates how many bytes of `result.result` actually contain randomness (required to support 32-bit engines, such as Mt19937).

Reply:
Oh, I see, thanks!

Reply:
The intent here was to ensure that the engine was indeed 64-bit, and that this invariant survives refactorings in zend_alloc or in ext/random.

Reply:
From the ext/random side, the behavior of xoshiro256** is guaranteed by https://github.com/php/php-src/blob/master/ext/random/tests/02_engine/xoshiro256starstar_value.phpt. If either the output size or implementation would change, that test would fail. That's the whole point of using a well-defined engine.

If you feel more safe with the assert in place, I won't object, just wanted to point out the guarantees of the randomness API for you to make an informed decision.

Reply:
```C
static php_random_result generate(void *state)
{
	return (php_random_result){
		.size = sizeof(uint64_t),
		.result = generate_state(state),
	};
}
```

it's indeed pretty obvious from the code, but I guess the assert doesn't hurt, if only to save a look into `ext/random/engine_xoshiro256starstar.c` to ensure that this PRNG is always returning `sizeof(uint64_t)` bytes of entropy.